src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -423,6 +429,88 @@ data CodingFailureMode
+
+    handleError table e s =
+        case cfm of
+            ErrorOnCodingFailure -> return $ ParserD.Error e
+            TransliterateCodingFailure ->
+                return $ case s of
+                    Just _ -> ParserD.Done 1 replacementChar
+                    Nothing -> ParserD.Continue 0 UTF8CharDecodeInit
+            DropOnCodingFailure ->
+                case s of
+                    Just x -> step' table UTF8CharDecodeInit x
+                    Nothing -> return $ ParserD.Partial 0 UTF8CharDecodeInit
+
+    {-# INLINE step' #-}
+    step' table UTF8CharDecodeInit x =
+        -- Note: It is important to use a ">" instead of a "<=" test
+        -- here for GHC to generate code layout for default branch
+        -- prediction for the common case. This is fragile and might
+        -- change with the compiler versions, we need a more reliable
+        -- "likely" primitive to control branch predication.
+        case x > 0x7f of
+            False -> return $ ParserD.Done 0 $ unsafeChr $ fromIntegral x
+            True ->
+                let (Tuple' sv cp) = decode0 table x
+                 in case sv of
+                        12 ->
+                            let msg = prefix
+                                    ++ "Invalid first UTF8 byte" ++ show x
+                             in handleError table msg Nothing
+                        0 -> return
+                                 $ ParserD.Error $ prefix ++ "unreachable state"
+                        _ -> return
+                                 $ ParserD.Continue 0 (UTF8CharDecoding sv cp)
+
+
+    step' table (UTF8CharDecoding statePtr codepointPtr) x =
+        let (Tuple' sv cp) = decode1 table statePtr codepointPtr x
+         in case sv of
+            0 -> return $ ParserD.Done 0 $ unsafeChr cp
+            12 ->
+                let msg = prefix
+                        ++ "Invalid subsequent UTF8 byte"
+                        ++ show x
+                        ++ "in state"
+                        ++ show statePtr
+                        ++ "accumulated value"
+                        ++ show codepointPtr
+                 in handleError table msg (Just x)
+            _ -> return $ ParserD.Continue 0 (UTF8CharDecoding sv cp)
+
+    {-# INLINE extract #-}
+    extract _ = error $ prefix ++ "Not enough input"
+
+-- XXX This should ideally accept a "CodingFailureMode" and perform appropriate
+-- error handling. This isn't possible now as "TransliterateCodingFailure"'s
+-- workflow requires backtracking 1 element. This can be revisited once "Fold"
+-- supports backtracking.
+{-# INLINE writeCharUtf8' #-}
+writeCharUtf8' :: MonadThrow m => Fold m Word8 Char
+writeCharUtf8' =  ParserD.toFold (parseCharUtf8WithD' ErrorOnCodingFailure)
+
+{-# INLINE parseCharUtf8' #-}
+parseCharUtf8' :: MonadCatch m => Parser.Parser m Word8 Char
+parseCharUtf8' = ParserK.toParserK (parseCharUtf8WithD' ErrorOnCodingFailure)
```

Comment:
In case, the alternative composition is not be very efficient we may have to pass the error option to this parser which is guaranteed to be efficient. You can try measuring the performance of alternative to decide.